### DISCUSSION TITLE
indentation-based blocks

### DISCUSSION THREAD

**Discussion by w1kee** [pest-parser/pest#716] [Q&A] [answered]
hi, i am making a parser and i want to have significant indentation in my language. i have seen the other posts about indentation, but i'm not sure how to do something like this.
example of language:
```
foo
    bar
```
i want to be able to parse this example like this:
```
file = { "foo" ~ startblock ~ "bar" ~ endblock }
...
```
is this possible, if yes, how?

**Comment by tomtau** [comment]
see e.g. https://github.com/thomas9911/python_parser/blob/master/src/grammar.pest#L32 ?

**Reply by KSXGitHub** [reply]
This doesn't work. It cannot parse two subblocks inside a parent block.

**Comment by alensiljak** [comment]
I think you need to explicitly mark the NEWLINE after "foo" and add some indentation before "bar".
I have something similar for Ledger postings [here](https://github.com/ledger-rs/ledger-pest-parser/blob/main/tests/def.pest#L32).

**Comment by simonbuchan** [comment, marked as answer]
This seems to work for me:

```pest
start = _ { SOI ~ block* ~ EOI }

block = { id ~ nl ~ body? }
body = _ { indent_start ~ block ~ (indent ~ block)* ~ indent_end }

indent = _ { PEEK[..] }
indent_start = _ { PEEK[..] ~ PUSH(" "+) }
indent_end = _ { DROP }

nl = _ { (" "* ~ NEWLINE)+ | EOI }
id = { ASCII_ALPHA+ }
```

With input:

```
a
  b
   c
  d
  
  e
     f
```

the editor gives:

```
- block
  - id: "a"
  - block
    - id: "b"
    - block > id: "c"
  - block > id: "d"
  - block
    - id: "e"
    - block
      - id: "f"
      - EOI: ""
- EOI: ""
```

(don't know why the share button doesn't work for me?)

**Reply by KSXGitHub** [reply]
What does `[..]` in `PEEK[..]` do?

**Reply by tristan-f-r** [reply]
It's covered in https://docs.rs/pest/latest/pest/#advanced-peeking - it reads the tabs bottom to top, as intended in an indentation schema.

**Reply by KSXGitHub** [reply]
Damn. Documentation is all over the place. I expected the pest book to contain all information I need.
